server/src/handler_driver.c

```diff
@@ -168,6 +192,412 @@ static bool drive_route_params_valid(const char *train_id, t_interlocking_route
+		
+		// C. Set t's trivial members to appropriate defaults (set to stop, is source, is destination)
+		info_arr.data_ptr[i]->has_been_set_to_stop = false;
+		info_arr.data_ptr[i]->is_source_signal = (i == 0);
+		info_arr.data_ptr[i]->is_destination_signal = (i+1 == number_of_signal_infos);
+		
+		// D. Allocate memory for non-trivial member 'id' in t
+		info_arr.data_ptr[i]->id = (char *) malloc(sizeof(char) * (strlen(signal_id_item) + 1));
+		if (info_arr.data_ptr[i]->id == NULL) {
+			syslog_server(LOG_ERR, "Building signal-infos: unable to allocate memory for "
+			              "new element ID");
+			free_route_signal_info_array(&info_arr);
+			return info_arr;
+		}
+		
+		// E. Copy ID from signal (from route->signals) into newly allocated 'id' of t
+		strcpy(info_arr.data_ptr[i]->id, signal_id_item);
+		
+		// F. Determine Index of the current signal in route->path and set member 
+		//    for that index of t accordingly
+		info_arr.data_ptr[i]->index_in_route_path = 0;
+		if (!info_arr.data_ptr[i]->is_source_signal && !info_arr.data_ptr[i]->is_destination_signal) {
+			for (size_t path_index = 0; path_index < route->path->len; ++path_index) {
+				const char *path_item = g_array_index(route->path, char *, path_index);
+				if (path_item != NULL && strcmp(path_item, signal_id_item) == 0) {
+					info_arr.data_ptr[i]->index_in_route_path = path_index;
+					break;
+				}
+			}
+		}
+		++built_signal_infos_counter;
+	}
+	return info_arr;
+}
+
+
+void free_multi_segment_indices(t_route_multi_segment_indices *d_seg_indices) {
+	if (d_seg_indices != NULL && d_seg_indices->arr != NULL) {
+		free(d_seg_indices->arr);
+		d_seg_indices->arr = NULL;
+		d_seg_indices->len = 0;
+	}
+}
+
+// "Multi segment index" = Index of a segment that appears more than once in a route
+static t_route_multi_segment_indices get_multi_segment_indices_for_route(const t_interlocking_route *route) {
+	// For route, determine which segments appear more than once
+	t_route_multi_segment_indices d_seg_indices = {.arr = NULL, .len = 0};
+	if (route == NULL || route->path == NULL) {
+		return d_seg_indices;
+	}
+	// First allocate enough memory for more than the maximum possible amount of 
+	// duplicate segment indices. Will afterwards be reallocated to the size it actually needs.
+	d_seg_indices.arr = malloc(sizeof(size_t) * route->path->len);
+	if (d_seg_indices.arr == NULL) {
+		syslog_server(LOG_ERR, 
+		              "Get multi segment indices for route %s: Unable to allocate memory", route->id);
+		return d_seg_indices;
+	}
+	// For every segment in route->path, check if it occurs again at a different position. 
+	// If yes, add both occurrences to d_seg_indices.
+	for (size_t path_index_i = 0; path_index_i < route->path->len; ++path_index_i) {
+		const char *path_item_i = g_array_index(route->path, char *, path_index_i);
+		if (path_item_i != NULL && is_type_segment(path_item_i)) {
+			for (size_t path_index_n = 0; path_index_n < route->path->len; ++path_index_n) {
```

Comment:
No, we still need to search the whole path - specifically because we only add the "outer" occurrence on line 365. If it would add both, we would only need to search the rest of the path indeed. My in-code comment suggesting to add both was outdated.